aws-android-sdk-iot-test/src/androidTest/java/com/amazonaws/mobileconnectors/iot/CustomAuthorization.java

```diff
@@ -0,0 +1,121 @@
+import com.amazonaws.util.Base64;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+
+final class CustomAuthorization {
+    private static final String CUSTOM_AUTHORIZER_NAME = "customAuthorizer";
+
+    private final String tokenKeyName;
+    private final String tokenValue;
+    private final String tokenSignature;
+    private final String customAuthorizerName;
+    private final String customAuthorizerLambdaArn;
+
+    private CustomAuthorization(
+            String tokenKeyName,
+            String tokenValue,
+            String tokenSignature,
+            String customAuthorizerName,
+            String customAuthorizerLambdaArn) {
+        this.tokenKeyName = tokenKeyName;
+        this.tokenValue = tokenValue;
+        this.tokenSignature = tokenSignature;
+        this.customAuthorizerName = customAuthorizerName;
+        this.customAuthorizerLambdaArn = customAuthorizerLambdaArn;
+    }
+
+    static CustomAuthorization from(JSONObject iotPackageConfiguration, KeysAndCertificateInfo keysAndCertificateInfo) {
+        String tokenKeyName = extractString(iotPackageConfiguration, ConfigKey.CUSTOM_AUTHORIZER_TOKEN_KEY_NAME);
+        String tokenValue = extractString(iotPackageConfiguration, ConfigKey.CUSTOM_AUTHORIZER_TOKEN_VALUE);
+        PrivateKeyReader reader = new PrivateKeyReader(keysAndCertificateInfo.getPrivateKey());
+        PrivateKey privateKey;
+        try {
+            privateKey = reader.getPrivateKey();
+        }  catch (IOException | InvalidKeySpecException parseFailure) {
+            throw new RuntimeException(parseFailure);
+        }
+        String tokenSignature = SHA256withRSASignature.sign(tokenValue, privateKey);
+        String customAuthorizerLambdaArn =
+            extractString(iotPackageConfiguration, ConfigKey.CUSTOM_AUTHORIZER_LAMBDA_ARN);
+
+        return new CustomAuthorization(
+            tokenKeyName,
+            tokenValue,
+            tokenSignature,
+            CUSTOM_AUTHORIZER_NAME,
+            customAuthorizerLambdaArn
+        );
+    }
+
+    private static String extractString(JSONObject configuration, ConfigKey key) {
+        try {
+            return configuration.getString(key.toString());
+        } catch (JSONException extractionFailure) {
+            throw new RuntimeException(extractionFailure);
+        }
+    }
+
+    String getTokenKeyName() {
+        return tokenKeyName;
+    }
+
+    String getTokenValue() {
+        return tokenValue;
+    }
+
+    String getTokenSignature() {
+        return tokenSignature;
+    }
+
+    public String getCustomAuthorizerName() {
```

Comment:
Sorry, dumb question maybe, is this visibility difference intentional?

Reply:
Nope! Not intentional. I've just been sloppy with them today. :-D Will fix.